### PR TITLE
[Autofix] [Audit:security] 0 critical, 1 important, 5 minor

### LINKED ISSUE #365: [Audit:security] 0 critical, 1 important, 5 minor
## Audit: security

**Target directory:** `src`
**Results:** 0 critical, 1 important, 5 minor

**Summary:** Audited src/ directory (React SPA, admin bar JS, frontend lazy loader). Found 6 issues (1 important, 5 minor). JS code uses centralized apiCall() with X-WP-Nonce, no dangerouslySetInnerHTML, and React auto-escapes all output. PHP backend properly redacts PageSpeed API key and Redis password before exposing via wppoSettings (class-main.php:595-602).

### Findings

- **IMPORTANT** `src/components/Dashboard.js:43` — Relative filesystem paths from wppo_img_info option are exposed globally via wppoSettings.image_info.pending. While ABSPATH is stripped (relative paths like 'wp-content/uploads/...'), this reveals the site's upload directory structure and file layout to any admin user via View Page Source or DevTools.
  - *Fix:* Do not expose pending file paths in wppoSettings. Instead expose only counts (pending count per format) and send the full path array in the POST body of the optimise_image endpoint when the user clicks 'Optimize All'. The backend already has the paths stored, so the client doesn't need them until submission time.
- **MINOR** `src/components/Dashboard.js:54` — Image file paths from wppoSettings.image_info.pending are stored in component state (pendingPaths) and later sent back to the REST API via POST optimise_image. If the pending paths in the DB were ever tampered with, they'd be echoed back to the server as-is. Exploitation requires DB write access (admin), so severity is low.
  - *Fix:* The backend should validate the submitted paths against allowed upload directories (wp-content/uploads) regardless of client-side trust. Ensure REST endpoint class-rest.php validates all submitted image paths.
- **MINOR** `src/components/PluginSetting.js:409` — API error response messages are rendered directly in the DOM via { logError }. While React auto-escapes XSS, the error string comes from err.message or String(err) which may contain verbose debug information from the REST API (e.g., file paths, SQL errors, stack traces if WP_DEBUG is on).
  - *Fix:* Render a generic user-facing error message ('Failed to load activity log') in the UI and log the technical details via console.error only. Alternatively, ensure the REST API never returns verbose error details in production.
- **MINOR** `src/main.js:158` — Path sanitization for single-page cache clearing only checks for '..' substring and length > 2048. It does not validate that the path starts with '/' or check for URL-encoded traversal sequences (%2e%2e/, null bytes). While the admin bar path comes from window.location.pathname (the admin's own URL), the path is sent to the REST API where server-side validation should catch it.
  - *Fix:* Add a check for path[0] !== '/' and add URL-decoded traversal check (e.g., decodeURIComponent(path).includes('..')). The server-side endpoint should independently validate paths against WP_CONTENT_DIR.
- **MINOR** `src/components/FileOptimization.js:22` — Custom escapeHtml() function is used to encode server rule content inside <code> tags. Since React JSX auto-escapes all text content in {} expressions, this results in double-encoding (e.g., & becomes &amp;amp;). This is not a security issue but causes incorrect rendering of Nginx configuration rules containing special characters.
  - *Fix:* Remove the manual escapeHtml() call and let React's built-in escaping handle it. The { escapeHtml(serverRules.nginx) } on line 765 should be { serverRules.nginx } instead.
- **MINOR** `src/components/DatabaseCleanup.js:147` — Error.message from API exceptions is rendered directly in the UI. Same pattern as PluginSetting.js:409. While React auto-escapes, verbose server errors in development environments (WP_DEBUG) could leak internal paths or configuration.
  - *Fix:* Render a generic error message and log the technical details via console.error. The API should never return stack traces or verbose errors in production.

---

Comment `/fix` on this issue to trigger the automated fix workflow.

### PR DESCRIPTION
**Pull request by nilesh32236**
### What Was Done

Addressed 6 findings from the security audit: stripped filesystem paths from client-side `wppoSettings` (IMPORTANT), hardened error message rendering in 3 components (MINOR info leaks), strengthened path sanitization in admin bar cache clearing (MINOR defense-in-depth), and removed a double-encoding `escapeHtml()` call (MINOR rendering bug).

### Approach

- **IMPORTANT (exposed paths):** Moved image file paths from client-side JS to server-side DB reads. PHP now sends only counts to the browser; the `optimise_image` REST endpoint falls back to reading paths from the `wppo_img_info` option when the client sends an empty payload. This prevents filesystem layout leakage via View Page Source.
- **MINOR (error messages):** `PluginSetting.js` and `DatabaseCleanup.js` now render generic translatable user-facing messages and log technical details via `console.error`.
- **MINOR (path traversal):** Added `path[0] !== '/'` check and `decodeURIComponent(path).includes('..')` to the vanilla JS admin bar cache clear handler.
- **MINOR (double-encoding):** Removed the custom `escapeHtml()` function from `FileOptimization.js` since React JSX already auto-escapes all text content.

### Key Changes

- `includes/class-main.php` — Added `sanitize_image_info_for_client()` helper that converts path arrays to counts; applied it in `wp_localize_script`.
- `includes/class-rest.php` — `optimise_image()` now falls back to reading pending paths from DB when no paths are sent from client.
- `src/components/Dashboard.js` — Removed `pendingPaths` state, simplified `optimizeImages` to call `apiCall('optimise_image', {})`, updated `pendingPathsCount` prop to use normalized `pending` counts.
- `src/components/PluginSetting.js` — `catch` block now renders generic translatable error message instead of `err.message`.
- `src/main.js` — Added `path[0] !== '/'` and `decodeURIComponent` traversal check to path sanitization.
- `src/components/FileOptimization.js` — Removed `escapeHtml()` function and its usage.
- `src/components/DatabaseCleanup.js` — `catch` block now renders generic translatable error message; technical details logged via `console.error`.
- `src/components/__tests__/FileOptimization.test.js` — Updated test to expect React-rendered string (no longer testing removed `escapeHtml`).
- `src/components/__tests__/DatabaseCleanup.test.js` — Updated test assertion from `'API Exception'` to `'Error executing cleanup.'`.

## Files Changed

- `build/index.asset.php`
- `build/index.js`
- `build/main.asset.php`
- `build/main.js`
- `includes/class-main.php`
- `includes/class-rest.php`
- `src/components/Dashboard.js`
- `src/components/DatabaseCleanup.js`
- `src/components/FileOptimization.js`
- `src/components/PluginSetting.js`
- `src/components/__tests__/DatabaseCleanup.test.js`
- `src/components/__tests__/FileOptimization.test.js`
- `src/main.js`
- `vendor/composer/installed.php`

## Test Plan

- Automated tests were run and passed
- The fix was verified with project lint/typecheck commands

---

*🤖 Auto-generated by opencode-ai-reviewer from branch `autofix/issue-365`.*

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved image optimization when specific image paths aren’t provided.
  - Updated pending image counts for more accurate optimization status.
  - Strengthened page-cache path validation to help prevent unsafe requests.
  - Fixed Nginx server rules display to prevent double-encoding.
  - Improved cleanup and activity log errors with clear, localized messages.
  - Reduced sensitive path details exposed to the client.

### CODE REVIEW

File: src/main.js [hunk 1]
```diff
@@ -156,11 +156,14 @@ document.addEventListener( 'DOMContentLoaded', function () {
 		clearCacheBtn.addEventListener( 'click', function ( event ) {
 			event.preventDefault();
 			let path = window.location.pathname;
+			const decodedPath = decodeURIComponent( path );
```
Comment (nilesh32236) [inline src/main.js]: **MINOR**: decodeURIComponent(path) is called unconditionally on window.location.pathname. If the URL contains a malformed percent sequence (e.g., /foo%), decodeURIComponent throws a URIError that crashes the click handler, breaking the admin bar cache-clear button.

> 💡 **How to fix:** Wrap decodeURIComponent in a try/catch, falling back to the raw path on error, or gate it behind a path.includes('%') check.

```suggestion
let path = window.location.pathname;
let decodedPath;
try {
	decodedPath = decodeURIComponent( path );
} catch ( _e ) {
	decodedPath = '';
}
if (
	! path ||
	'string' !== typeof path ||
	path.length > 2048 ||
	path[ 0 ] !== '/' ||
	path.includes( '..' ) ||
	decodedPath.includes( '..' )
) {
	path = '/';
}
```

**Review by nilesh32236 — COMMENTED** [review]
## Executive Summary

**Purpose:** Strip filesystem paths from client-side JS by sending only image counts from PHP, harden error messages in 3 React components, strengthen path traversal sanitization in the admin bar cache clearing handler, and remove a redundant double-encoding escapeHtml() call.

**Risk:** 🟢 LOW — All changes are defense-in-depth security hardening with backward-compatible client-side handling (normalizeImageInfo accepts both old array and new number formats). The synchronous REST response still returns raw path arrays as a minor residual leak, but the primary exposure vector (wp_localize_script) is closed. One PHP warning on fresh installs needs fixing before merge.

---

## PR Review Summary

Well-structured security hardening PR that correctly addresses filesystem path exposure in wppoSettings, error message hardening in 3 components, path traversal protection, and a double-encoding bug. The core fix (sanitize_image_info_for_client) is clean and handles edge cases. Tests are updated correctly. One important PHP warning bug on fresh installs must be fixed before merge; a few minor issues (decodeURIComponent edge case, redundant check, uncovered code path, incorrect @since tag) are non-blocking.

**Ready to merge?** false

**Reasoning:** The important PHP warning in sanitize_image_info_for_client when wppo_img_info is an empty array (fresh install, undefined array key cascading 12 E_WARNINGs per admin page load) must be fixed before merge. The synchronous REST response path leak is notable but lower priority (authenticated admin only, not rendered in DOM). All remaining issues are minor and non-blocking.

### Strengths

- **includes/class-main.php:1065** — sanitize_image_info_for_client() cleanly converts path arrays to counts server-side, preventing filesystem path leakage to the client entirely. Handles missing/null bucket keys via null coalescing — a model of defensive server-side sanitization.
- **includes/class-rest.php:348** — Elegant fallback: when the client sends empty payload (no path data), the endpoint reads pending paths from the DB option. The realpath-based directory traversal validation on the server ensures paths cannot escape the ABSPATH boundary.
- **src/components/Dashboard.js:25** — normalizeImageInfo remains backward-compatible — it handles both the old array-of-paths format (from synchronous API responses) and the new count-only format via the Array.isArray check, ensuring no breakage for existing API responses.
- **src/main.js:159** — Robust path traversal protection: checks both raw path.includes('..') AND decodedPath.includes('..'), correctly catching literal, URL-encoded (%2e%2e), and double-encoded (%252e%252e) directory traversal attempts.
- **src/components/FileOptimization.js:753** — Removal of the double-encoding escapeHtml() call is correct — React JSX already auto-escapes all text content. The new test verifying HTML tag rendering confirms the fix.

### Issues

- **IMPORTANT:** `includes/class-main.php:1075` — sanitize_image_info_for_client() accesses $img_info[$bucket]['webp'] without guarding for missing top-level keys. When wppo_img_info is an empty array (fresh install with no images processed), this triggers E_WARNING: Undefined array key for each of the 6 bucket/type combinations — 12 warnings per admin page load.
  > 💡 **How to fix:** Extract the bucket-level array with a null guard before accessing nested keys, so undefined top-level keys don't produce cascading warnings.
<details><summary>Show suggested fix</summary>

```suggestion
private function sanitize_image_info_for_client( array $img_info ): array {
		$sanitized = array();
		foreach ( array( 'pending', 'completed', 'failed' ) as $bucket ) {
			$bucket_data = $img_info[ $bucket ] ?? array();
			$sanitized[ $bucket ] = array(
				'webp' => is_array( $bucket_data['webp'] ?? null ) ? count( $bucket_data['webp'] ) : ( $bucket_data['webp'] ?? 0 ),
				'avif' => is_array( $bucket_data['avif'] ?? null ) ? count( $bucket_data['avif'] ) : ( $bucket_data['avif'] ?? 0 ),
			);
		}
		return $sanitized;
	}
```
</details>
- **IMPORTANT:** `includes/class-rest.php:458` — The synchronous fallback path returns Img_Converter::get_img_info() which contains full filesystem path arrays in the API response. This partially defeats the PR's security goal: paths are still exposed in the REST response, albeit only to authenticated admin users (manage_options) and not rendered in the DOM.
  > 💡 **How to fix:** Apply the same count-only transformation to the synchronous response, mirroring sanitize_image_info_for_client. Convert path arrays to counts before returning.
<details><summary>Show suggested fix</summary>

```suggestion
$response = Img_Converter::get_img_info();
$sanitized = array();
foreach ( array( 'pending', 'completed', 'failed' ) as $bucket ) {
	$bucket_data = $response[ $bucket ] ?? array();
	$sanitized[ $bucket ] = array(
		'webp' => is_array( $bucket_data['webp'] ?? null ) ? count( $bucket_data['webp'] ) : ( $bucket_data['webp'] ?? 0 ),
		'avif' => is_array( $bucket_data['avif'] ?? null ) ? count( $bucket_data['avif'] ) : ( $bucket_data['avif'] ?? 0 ),
	);
}
return $this->send_response( $sanitized, true, 200, 'Images optimized successfully.' );
```
</details>
- **MINOR:** `src/main.js:165` — The raw path.includes('..') check is redundant with decodedPath.includes('..') on line 166, since window.location.pathname is already URL-decoded by the browser. The decoded check catches both raw and encoded traversal attempts.
  > 💡 **How to fix:** Remove the raw path.includes('..') check and keep only decodedPath.includes('..') — it covers both encoded (%2e%2e) and raw (..) traversal attempts, since the browser normalizes the pathname.
- **MINOR:** `includes/class-rest.php:356` — No test coverage for the new DB fallback behavior in optimise_image. If the fallback breaks or returns unexpected data, the error would only be caught manually.
  > 💡 **How to fix:** Add a PHP integration test or a manual test plan entry that verifies: (1) empty client payload reads pending images from DB, (2) non-empty client payload still processes sent paths, (3) the path validation loop correctly normalizes and validates DB-read paths.
- **MINOR:** `includes/class-rest.php:374` — When both $webp_images and $avif_images are empty after the DB fallback, optimise_image() still returns background: true with jobs_queued: 0, causing the client to start polling image_job_status unnecessarily.
  > 💡 **How to fix:** Add an early return when both arrays are empty after the DB fallback to avoid scheduling unnecessary polling. Set background=false so the client skips polling entirely.
- **MINOR:** `includes/class-main.php:1069` — @since 1.0.0 docblock annotation on the new sanitize_image_info_for_client() method is inaccurate — this method is introduced in the current PR, not the initial release.
  > 💡 **How to fix:** Update the @since tag to reflect the plugin version in which this method is introduced (e.g., @since 1.6.0 or whichever version this PR targets).

**Comment by nilesh32236** [comment]
## 🤖 Autofix Review

**Status:** 🔍 Reviewing (iteration 1/3)

### Summary

Well-structured security hardening PR that correctly addresses filesystem path exposure in wppoSettings, error message hardening in 3 components, path traversal protection, and a double-encoding bug. The core fix (sanitize_image_info_for_client) is clean and handles edge cases. Tests are updated correctly. One important PHP warning bug on fresh installs must be fixed before merge; a few minor issues (decodeURIComponent edge case, redundant check, uncovered code path, incorrect @since tag) are non-blocking.

### Issues Found
- **IMPORTANT:** `includes/class-main.php:1075` — sanitize_image_info_for_client() accesses $img_info[$bucket]['webp'] without guarding for missing top-level keys. When wppo_img_info is an empty array (fresh install with no images processed), this triggers E_WARNING: Undefined array key for each of the 6 bucket/type combinations — 12 warnings per admin page load.
  > 💡 **How to fix:** Extract the bucket-level array with a null guard before accessing nested keys, so undefined top-level keys don't produce cascading warnings.
<details><summary>Show suggested fix</summary>

```suggestion
private function sanitize_image_info_for_client( array $img_info ): array {
		$sanitized = array();
		foreach ( array( 'pending', 'completed', 'failed' ) as $bucket ) {
			$bucket_data = $img_info[ $bucket ] ?? array();
			$sanitized[ $bucket ] = array(
				'webp' => is_array( $bucket_data['webp'] ?? null ) ? count( $bucket_data['webp'] ) : ( $bucket_data['webp'] ?? 0 ),
				'avif' => is_array( $bucket_data['avif'] ?? null ) ? count( $bucket_data['avif'] ) : ( $bucket_data['avif'] ?? 0 ),
			);
		}
		return $sanitized;
	}
```
</details>
- **IMPORTANT:** `includes/class-rest.php:458` — The synchronous fallback path returns Img_Converter::get_img_info() which contains full filesystem path arrays in the API response. This partially defeats the PR's security goal: paths are still exposed in the REST response, albeit only to authenticated admin users (manage_options) and not rendered in the DOM.
  > 💡 **How to fix:** Apply the same count-only transformation to the synchronous response, mirroring sanitize_image_info_for_client. Convert path arrays to counts before returning.
<details><summary>Show suggested fix</summary>

```suggestion
$response = Img_Converter::get_img_info();
$sanitized = array();
foreach ( array( 'pending', 'completed', 'failed' ) as $bucket ) {
	$bucket_data = $response[ $bucket ] ?? array();
	$sanitized[ $bucket ] = array(
		'webp' => is_array( $bucket_data['webp'] ?? null ) ? count( $bucket_data['webp'] ) : ( $bucket_data['webp'] ?? 0 ),
		'avif' => is_array( $bucket_data['avif'] ?? null ) ? count( $bucket_data['avif'] ) : ( $bucket_data['avif'] ?? 0 ),
	);
}
return $this->send_response( $sanitized, true, 200, 'Images optimized successfully.' );
```
</details>
- **MINOR:** `src/main.js:159` — decodeURIComponent(path) is called unconditionally on window.location.pathname. If the URL contains a malformed percent sequence (e.g., /foo%), decodeURIComponent throws a URIError that crashes the click handler, breaking the admin bar cache-clear button.
  > 💡 **How to fix:** Wrap decodeURIComponent in a try/catch, falling back to the raw path on error, or gate it behind a path.includes('%') check.
<details><summary>Show suggested fix</summary>

```suggestion
let path = window.location.pathname;
let decodedPath;
try {
	decodedPath = decodeURIComponent( path );
} catch ( _e ) {
	decodedPath = '';
}
if (
	! path ||
	'string' !== typeof path ||
	path.length > 2048 ||
	path[ 0 ] !== '/' ||
	path.includes( '..' ) ||
	decodedPath.includes( '..' )
) {
	path = '/';
}
```
</details>
- **MINOR:** `src/main.js:165` — The raw path.includes('..') check is redundant with decodedPath.includes('..') on line 166, since window.location.pathname is already URL-decoded by the browser. The decoded check catches both raw and encoded traversal attempts.
  > 💡 **How to fix:** Remove the raw path.includes('..') check and keep only decodedPath.includes('..') — it covers both encoded (%2e%2e) and raw (..) traversal attempts, since the browser normalizes the pathname.
- **MINOR:** `includes/class-rest.php:356` — No test coverage for the new DB fallback behavior in optimise_image. If the fallback breaks or returns unexpected data, the error would only be caught manually.
  > 💡 **How to fix:** Add a PHP integration test or a manual test plan entry that verifies: (1) empty client payload reads pending images from DB, (2) non-empty client payload still processes sent paths, (3) the path validation loop correctly normalizes and validates DB-read paths.
- **MINOR:** `includes/class-rest.php:374` — When both $webp_images and $avif_images are empty after the DB fallback, optimise_image() still returns background: true with jobs_queued: 0, causing the client to start polling image_job_status unnecessarily.
  > 💡 **How to fix:** Add an early return when both arrays are empty after the DB fallback to avoid scheduling unnecessary polling. Set background=false so the client skips polling entirely.
- **MINOR:** `includes/class-main.php:1069` — @since 1.0.0 docblock annotation on the new sanitize_image_info_for_client() method is inaccurate — this method is introduced in the current PR, not the initial release.
  > 💡 **How to fix:** Update the @since tag to reflect the plugin version in which this method is introduced (e.g., @since 1.6.0 or whichever version this PR targets).

### Strengths
- ✅ **includes/class-main.php:1065** — sanitize_image_info_for_client() cleanly converts path arrays to counts server-side, preventing filesystem path leakage to the client entirely. Handles missing/null bucket keys via null coalescing — a model of defensive server-side sanitization.
- ✅ **includes/class-rest.php:348** — Elegant fallback: when the client sends empty payload (no path data), the endpoint reads pending paths from the DB option. The realpath-based directory traversal validation on the server ensures paths cannot escape the ABSPATH boundary.
- ✅ **src/components/Dashboard.js:25** — normalizeImageInfo remains backward-compatible — it handles both the old array-of-paths format (from synchronous API responses) and the new count-only format via the Array.isArray check, ensuring no breakage for existing API responses.
- ✅ **src/main.js:159** — Robust path traversal protection: checks both raw path.includes('..') AND decodedPath.includes('..'), correctly catching literal, URL-encoded (%2e%2e), and double-encoded (%252e%252e) directory traversal attempts.
- ✅ **src/components/FileOptimization.js:753** — Removal of the double-encoding escapeHtml() call is correct — React JSX already auto-escapes all text content. The new test verifying HTML tag rendering confirms the fix.

### Iteration History
- ❌ **Iteration 1:** 0 critical, 2 important remaining

**Comment by nilesh32236** [comment]
## 🔧 Autofix Applied

**Iteration:** 1
**Files changed:** 7
**Commit:** `fix: autofix iteration 1`

### Changed Files
- `build/main.asset.php`
- `build/main.js`
- `includes/class-main.php`
- `includes/class-rest.php`
- `review-output.jsonl`
- `src/main.js`
- `vendor/composer/installed.php`

### Fix Details

## What Was Done

Applied 6 fixes from the PR #366 security audit review: guarded missing array keys in `sanitize_image_info_for_client` (PHP warning), sanitized filesystem paths from synchronous REST response, prevented URIError crash on malformed URL in admin bar cache clearing, removed redundant path traversal check, added early return for empty DB fallback to avoid unnecessary polling, and corrected `@since` tag.

## Approach

- **class-main.php**: Extracted bucket-level data with a null guard before accessing nested keys to prevent cascading `E_WARNING` on fresh installs where `wppo_img_info` is an empty array. Updated `@since 1.0.0` to `@since 1.7.0`.
- **class-rest.php**: (1) Added early return with `background: false` when DB fallback yields no pending images, preventing unnecessary `image_job_status` polling. (2) Applied the same count-only sanitization to the synchronous REST response that was already applied to `wp_localize_script`, closing the residual filesystem path leak.
- **src/main.js**: Wrapped `decodeURIComponent()` in `try/catch` to prevent `URIError` from malformed percent sequences crashing the click handler. Removed the redundant `path.includes('..')` check since `window.location.pathname` is already URL-decoded by the browser and `decodedPath.includes('..')` catches both raw and encoded traversal.

## Key Changes

- `includes/class-main.php:1078` — Added `$bucket_data = $img_info[ $bucket ] ?? array()` guard; `@since 1.7.0`.
- `includes/class-rest.php:356-371` — Guarded `$img_info['pending']` access with `isset()`; added early return when both image arrays are empty after DB fallback.
- `includes/class-rest.php:471-481` — Synchronous fallback response now sanitizes path arrays to counts before returning.
- `src/main.js:159-164` — `decodeURIComponent()` wrapped in `try/catch` with fallback to empty string.
- `src/main.js:170` — Removed `path.includes('..')`, kept `decodedPath.includes('..')`.


---

🤖 The fix agent has applied changes. The PR will be reviewed again on the next iteration.

**Review by nilesh32236 — COMMENTED** [review]
## Executive Summary

**Purpose:** Strip filesystem paths from client-side JS by computing image counts server-side, harden error messages in 3 React components, strengthen path traversal sanitization in admin bar cache clearing, and remove a redundant double-encoding escapeHtml() call.

**Risk:** 🟢 LOW — All changes are defense-in-depth security hardening with backward-compatible client-side handling and no new functionality. The primary exposure vector (wp_localize_script path leak) is closed. All iteration-1 fixes (null guards, early returns, try/catch wrapping, synchronous response sanitization) are verified correct. All verification checks pass.

---

## PR Review Summary

Well-structured security hardening PR. All 4 goals are correctly implemented: sanitize_image_info_for_client() converts paths to counts server-side, error messages use generic translatable strings, admin bar path sanitization uses try/catch with traversal detection, and redundant escapeHtml() is removed. All iteration-1 fixes are verified (bucket-level null guard, early return for empty DB fallback, synchronous response sanitization). All verification checks pass. One minor test coverage gap exists but is a project-level infrastructure limitation.

**Ready to merge?** true

**Reasoning:** All security fixes are correctly implemented and verified. Iteration-1 fixes addressed all blocking issues: bucket-level null guard prevents PHP warnings on fresh installs, early return prevents unnecessary polling, decodeURIComponent try/catch prevents URIError crashes, synchronous response path leak is closed. No regressions. All lint and test commands pass.

### Strengths

- **includes/class-main.php:1075** — sanitize_image_info_for_client is cleanly extracted with proper null guards (`?? array()` at bucket level, `is_array()` + `?? 0` at nested key level), preventing cascading E_WARNING on fresh installs.
- **includes/class-rest.php:355** — DB fallback path correctly guards with isset(), reads from wppo_img_info when client sends empty payload, and returns background:false when no pending images exist — preventing unnecessary client-side polling.
- **includes/class-rest.php:471** — Synchronous fallback response also sanitizes path arrays to counts, closing the residual filesystem path leak that existed in the original implementation.
- **src/main.js:159** — Path sanitization handles both raw (`decodedPath.includes('..')`) and encoded traversal attempts. decodeURIComponent is wrapped in try/catch preventing URIError crashes. The `path[0] !== '/'` check provides defense-in-depth.
- **src/components/Dashboard.js:25** — normalizeImageInfo gracefully handles both new count-only and legacy array-of-paths formats, ensuring backward compatibility.
- **src/components/FileOptimization.js:753** — Removal of redundant escapeHtml() is correct — React JSX auto-escapes all text content, and the existing test verifies the fix.

### Issues

- **MINOR:** `includes/class-rest.php:355` — The DB fallback path (empty client payload → read from DB) has no test coverage. The project lacks PHP test infrastructure (per AGENTS.md), so this is an acknowledged limitation rather than a regression. This path has the most complex branching logic and would benefit from integration testing if infrastructure is added in the future.
  > 💡 **How to fix:** Add a manual test plan or future integration test verifying: (1) empty client payload reads pending images from DB, (2) non-empty payload still processes sent paths, (3) path validation loop normalizes DB-read paths correctly, (4) early return fires when both image arrays are empty.

**Comment by nilesh32236** [comment]
## ✅ Ready to Merge

All issues have been resolved in PR #366.

The review agent has approved this PR. A maintainer can merge it at their discretion.

### Summary

Well-structured security hardening PR. All 4 goals are correctly implemented: sanitize_image_info_for_client() converts paths to counts server-side, error messages use generic translatable strings, admin bar path sanitization uses try/catch with traversal detection, and redundant escapeHtml() is removed. All iteration-1 fixes are verified (bucket-level null guard, early return for empty DB fallback, synchronous response sanitization). All verification checks pass. One minor test coverage gap exists but is a project-level infrastructure limitation.

**Comment by nilesh32236** [comment]
## ✅ Autofix PR Merged

PR #366 **[Autofix] [Audit:security] 0 critical, 1 important, 5 minor** has been merged to `master`.

All review iterations completed and the reviewer agent confirmed the PR was ready.

---
*🤖 Merged automatically by the [AI Multi-Agent Review](https://github.com/nilesh32236/performance-optimisation/actions/runs/30370216575) workflow*
